### PR TITLE
Pr3 column visibility persistance

### PR DESCRIPTION
**Pull request by GALTdea**
### What github issue is this PR for, if any?
This commit is the 3rd and last Pr to fix issue #2698 

### What changed, and why?
In this Pr I add client side code to allow users to persist column visibility status. This PR adds: 
PreferencesSets controller
- PreferencesSetTableStateService: This is a service object that is incharge of accessing and manipulating table states and handles potential errors during the update process. 
-Add PreferenceSets route resources
-Add Preference_set_table_state_service_spec to test PreferencesSetTableStateService 
-Adds new settings  VoluntersTable in dashbjoar.js. These new changes will allow the new table state to be persisted to the appropriate database table.

### How will this affect user permissions?
- Volunteer permissions:n/a
- Supervisor permissions:n/a
- Admin permissions:n/a

### How is this tested? (please write tests!) 💖💪
I added preference_set_table_state_service_spec.rb to ensure that  table_state is being loaded and updated correctly.

### Feelings gif (optional)
What gif best describes your feeling working on this issue? https://giphy.com/
How to embed:
`alt text`

### Feedback please? (optional)
We are very interested in your feedback! Please give us some :) https://forms.gle/1D5ACNgTs2u9gSdh9

### CODE REVIEW

**Review by littleforest — COMMENTED** [review]
@GALTdea thanks so much for working on this! The stakeholders are going to be so excited. I tried it out in my dev environment, and I can see the `preference_set` record getting updated, but the checkboxes aren't persisting when I leave the page and come back, or log out and log back in again 🤔 . Could you add some system specs that test the behavior as well?

File: spec/requests/preference_sets_spec.rb
[the remark this answers is not in the document]
Comment (GALTdea) [reply spec/requests/preference_sets_spec.rb:1]: Hi @littleforest, thank you so much for reviewing my PR so quickly! I'm happy to double-check my code and create that system test.

**Comment by GALTdea** [comment]
> @GALTdea thanks so much for working on this! The stakeholders are going to be so excited. I tried it out in my dev environment, and I can see the `preference_set` record getting updated, but the checkboxes aren't persisting when I leave the page and come back, or log out and log back in again 🤔 . Could you add some system specs that test the behavior as well?

This has been a great learning experience for me. I was happy to learn abut datatables. I thinks it's a pretty neat library

**Comment by GALTdea** [comment]
Hi @littleforest @compwron, i just completed request test preference_sets_spec. [https://github.com/GALTdea/casa/blob/pr3_column_visibility_persistance/spec/requests/preference_sets_spec.rb](url) I had also completed a service test to test the preference_set_table_state_service. https://github.com/GALTdea/casa/blob/pr3_column_visibility_persistance/spec/services/preference_set_table_state_service_spec.rb hope this helps.

**Comment by GALTdea** [comment]
Also, I don't think the failing tests are within the scope of this issue.

**Comment by compwron** [comment]
The spec checker says you're missing a test, and also a lot of rspec tests are failing... We can't merge this as is but maybe we can rebase it and dig in to find a fix. @littleforest @FireLemons maybe at the next office hours we can land this? I can try to take a look but I'm not sure when

**Comment by GALTdea** [comment]
Than you @compwron  yes, I saw that I was missing that test. Also I see that a lot of test are failing. locally I'm only failing two tests. I can try to make it to the office hours.

File: spec/requests/preference_sets_spec.rb
[the remark this answers is not in the document]
Comment (FireLemons) [reply spec/requests/preference_sets_spec.rb:1]: I checked out the pr @littleforest and got the feature to work. Did you remember to recompile js assets?

**Comment by GALTdea** [comment]
Hi @compwron and @littleforest sorry it took me such a long time to add the System test requested. I finally had a chance to complete it so I would appreciate a review. Just as a refresher. I paired with @compwron during office hours a few weeks ago. He verified that the column visibility settings were in fact persisting. He asked me to complete a system test for it.

**Comment by compwron** [comment]
very cool :) 
I think that during office hours your probably paired with @FireLemons not me :)

**Review by compwron — APPROVED** [review]

**Comment by GALTdea** [comment]
Thank you for merging!  And I'm so sorry! Yes, I met with @FireLemons!

**Pull request merged**